File: Weekly-402/Q4.py
class Solution:
    
    def countOfPeaks(self, nums: List[int], queries: List[List[int]]) -> List[int]:
        n = len(nums)
        u = 1
        while u <= n: 
            u *= 2
        seg = [0] * (u * 2)
        
        def build(v, tl, tr):
            if tl == tr:
                if tl - 1 >= 0 and tl + 1 < n:
                    seg[v] = 1 if nums[tl] > nums[tl - 1] and nums[tl] > nums[tl + 1] else 0
                else:
                    seg[v] = 0
            else:
                tm = (tl + tr) // 2
                build(2 * v, tl, tm)
                build(2 * v + 1, tm + 1, tr)
                seg[v] = seg[2 * v] + seg[2 * v + 1]
        
        def get(v, tl, tr, l, r):
            if l > r:
                return 0
            if tl == l and tr == r:
                return seg[v]
            tm = (tl + tr) // 2
            return get(2 * v, tl, tm, l, min(tm, r)) + get(2 * v + 1, tm + 1, tr, max(tm + 1, l), r)
        
        def update(v, tl, tr, pos, val):
            if tl == tr:
                nums[tl] = val
                if tl - 1 >= 0 and tl + 1 < n:
                    seg[v] = 1 if nums[tl] > nums[tl - 1] and nums[tl] > nums[tl + 1] else 0
                else:
                    seg[v] = 0
                
            else:
                tm = (tl + tr) // 2
                if pos <= tm:
                    update(2 * v, tl, tm, pos, val)
                else:
                    update(2 * v + 1, tm + 1, tr, pos, val)
                seg[v] = seg[2 * v] + seg[2 * v + 1]
        
        
        build(1, 0, n - 1)
        ans = []
        for x in queries:
            if x[0] == 1:
                ans.append(get(1, 0, n - 1, x[1]+1, x[2]-1))
            else:
                update(1, 0, n - 1, x[1], x[2])
                if x[1]-1>=0:
                    update(1, 0, n-1, x[1]-1, nums[x[1]-1])
                if x[1]+1<n:
                    update(1, 0, n-1, x[1]+1, nums[x[1]+1])
        
        return ans

File: Weekly-407/Q4.cpp
class Solution {
public:
    long long solve(vector<int> &x) {
        long long ans = abs(x[0]);
        for(int i=1;i<x.size();i++){
            ans+=max(abs(x[i])-abs(x[i-1]),0);
        }
        return ans;
    }
    long long minimumOperations(vector<int>& nums, vector<int>& target) {
        int n = nums.size();
        vector<int> diff(n);
        for(int i=0; i<n; i++)diff[i]=nums[i]-target[i];
        
        vector<vector<int>> neg, pos;
        vector<int> tmp;
        
        for(int i=0; i<n; i++) {
            if(diff[i]<0) {
                tmp.clear();
                while(i<n&&diff[i]<0) tmp.push_back(diff[i++]);
                neg.push_back(tmp);
                i--;
            }else if(diff[i]>0){
                tmp.clear();
                while(i<n&&diff[i]>0) tmp.push_back(diff[i++]);
                pos.push_back(tmp);
                i--;
            }
        }
        long long tot=0;
        for(auto x: pos) {
            // for(auto c: x)cout<<c<<' ';cout<<'\n';
            tot+=solve(x);
        }
        for(auto y: neg){
            // for(auto c: y)cout<<c<<' ';cout<<'\n';
            tot+=solve(y);
        }
        return tot;
        
        
    }
};
